lib/services/local/saml_idp_service_provider.go

```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
 	sp.SetEntityDescriptor(string(entityDescriptor))
 	return nil
 }
+
+// embedAttributeMapping embeds attribute mapping details into entity descriptor.
+func (s *SAMLIdPServiceProviderService) embedAttributeMapping(sp types.SAMLIdPServiceProvider) error {
+	if len(sp.GetAttributeMapping()) == 0 {
+		s.log.Infof("no custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
+		return nil
+	}
+
+	ed, err := samlsp.ParseMetadata([]byte(sp.GetEntityDescriptor()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// return if provided attributes are already embedded.
+	embeddedSPSSODescriptorIndex, embeddedAttributes := samlSPSSODescriptorToProtoSAMLAttributeMapping(ed.SPSSODescriptors)
+	if reflect.DeepEqual(sp.GetAttributeMapping(), embeddedAttributes) {
+		return nil
+	}
```

Comment:
I've a couple of questions about this block:
- In what case will the attributes be already embedded?
- Is "deep equal" check appropriate here? What if the attribute mapping already has a superset of attributes?

Reply:
> In what case will the attributes be already embedded?

example:
1. user creates sp with entity descriptor and attribute mapping. initial sets of attributes are embedded. 
2. Then, user wants to update attribute, does `tctl get saml_idp_service_provider/sp > sp.yml`, adds new attributes to `sp.yml` and then updates. At this point, since there will be a set of attributes already embedded in entity descriptor (from step 1). If we do not check for that, we risk duplicate them. (that's why go through loop to find index)

> Is "deep equal" check appropriate here? What if the attribute mapping already has a superset of attributes?

So while I was working, I noticed `CreateSAMLIdPServiceProvider` and `UpdateSAMLIdPServiceProvider` were invoked more times than I initially would have thought (during backend insert, during cache upsert) so I thought if it would be good to break out from embedding if entity descriptor already has all the attributes mapped, hence used deep equal here to compare whats in attribute mapping vs. entity descriptor.

Reply:
fyi, I have updated this and now excludes `reflect.DeepEqual` and only return early if attribute mapping input and teleport managed entity descriptor is empty. It also now handles attribute mapping deletion from the entity descriptor if it is removed from the `saml_idp_service_provider` spec, which was not handled previously.